trieste/acquisition/rule.py

```diff
@@ -37,86 +38,73 @@
     SingleModelAcquisitionBuilder,
     SingleModelGreedyAcquisitionBuilder,
 )
-from .optimizer import (
-    AcquisitionOptimizer,
-    automatic_optimizer_selector,
-    batchify,
-    generate_continuous_optimizer,
-)
+from .optimizer import AcquisitionOptimizer, automatic_optimizer_selector, batchify
 from .sampler import ExactThompsonSampler, RandomFourierFeatureThompsonSampler, ThompsonSampler
 
-S = TypeVar("S")
-""" Unbound type variable. """
+T_co = TypeVar("T_co", covariant=True)
+""" Unbound covariant type variable. """
 
 SP_contra = TypeVar("SP_contra", bound=SearchSpace, contravariant=True)
 """ Contravariant type variable bound to :class:`~trieste.space.SearchSpace`. """
 
 
-class AcquisitionRule(ABC, Generic[S, SP_contra]):
-    """The central component of the acquisition API."""
+class AcquisitionRule(ABC, Generic[T_co, SP_contra]):
+    """
+    The central component of the acquisition API.
+
+    An :class:`AcquisitionRule` can produce any value from the search space for this step, and the
```

Comment:
The main thing I'm not certain about in this PR is that this definition suggests that `T_co` can be anything, but in practice it has to be either `TensorType` or `State[Optional[State], TensorType]` if we want to be able to use it in the BO loop. Should we state this explicitly in the definition of `T_co`?

Reply:
I don't know of any way we could type it, since a `TypeVar` can't contain another `TypeVar` and `State` has to be invariant in its state type. Did you have something in mind?

> in practice it has to be `TensorType` or `State[Optional[State], TensorType]`

In the `BayesianOptimizer` that's true, but users, including manchester, might not be using that

Reply:
Ok. Then maybe just change this comment to be a bit more assertive? Something like: `To use this in the Bayesian Optimization loop, this value needs to be a set of query points, either on its own...`

Reply:
how about appending
```
    Indeed, to use an :class:`AcquisitionRule` in the main
    :class:`~trieste.bayesian_optimizer.BayesianOptimizer` Bayesian optimization loop, the rule
    must return either a `TensorType` or `State`-ful `TensorType`.
```